File: spikes/language/ts-bun/src/parse.ts
import { readFileSync } from "fs";
import { basename, extname } from "path";

// Use Bun's built-in tree-sitter support if available, otherwise fall back to node-tree-sitter
// Note: Bun does not have built-in tree-sitter. We use the node binding.

const Parser = (await import("tree-sitter")).default;
const LangRust = (await import("tree-sitter-rust")).default;
const LangJS = (await import("tree-sitter-javascript")).default;
const LangPython = (await import("tree-sitter-python")).default;

const filePath = process.argv[2];
const iterations = parseInt(process.argv[3] || "10", 10);

if (!filePath) {
  console.error("Usage: bun run src/parse.ts <file-path> [iterations]");
  process.exit(1);
}

const source = readFileSync(filePath, "utf-8");
const lineCount = source.split("\n").length;

const ext = extname(filePath).slice(1);
const parser = new Parser();

const langMap: Record<string, any> = {
  rs: LangRust,
  js: LangJS,
  ts: LangJS, // JS parser handles TS syntax partially
  jsx: LangJS,
  tsx: LangJS,
  py: LangPython,
};

const lang = langMap[ext];
if (!lang) {
  console.error(`Unsupported extension: ${ext}`);
  process.exit(1);
}
parser.setLanguage(lang);

// Warmup parse
parser.parse(source);

// Timed iterations
const start = performance.now();
let nodeCount = 0;

for (let i = 0; i < iterations; i++) {
  const tree = parser.parse(source);
  // Walk all nodes to ensure full parse
  const cursor = tree.walk();
  let reachedRoot = false;
  while (!reachedRoot) {
    nodeCount++;
    if (cursor.gotoFirstChild()) continue;
    if (cursor.gotoNextSibling()) continue;
    let retracing = true;
    while (retracing) {
      if (!cursor.gotoParent()) {
        retracing = false;
        reachedRoot = true;
      } else if (cursor.gotoNextSibling()) {
        retracing = false;
      }
    }
  }
}

const elapsed = performance.now() - start;
const avgMs = elapsed / iterations;

console.log(
  JSON.stringify(
    {
      tool: "ts-bun",
      operation: "tree_sitter_parse",
      file: filePath,
      lines: lineCount,
      bytes: Buffer.byteLength(source, "utf-8"),
      iterations,
      total_ms: parseFloat(elapsed.toFixed(2)),
      avg_ms: parseFloat(avgMs.toFixed(3)),
      nodes_per_iter: Math.round(nodeCount / iterations),
    },
    null,
    2
  )
);

File: spikes/language/rust/src/parse.rs
use std::env;
use std::fs;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: spike-parse <file-path> [iterations]");
        std::process::exit(1);
    }
    let file_path = &args[1];
    let iterations: u32 = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(10);

    let source = fs::read_to_string(file_path).expect("Failed to read file");
    let line_count = source.lines().count();

    // Detect language from extension
    let ext = file_path.rsplit('.').next().unwrap_or("");
    let mut parser = tree_sitter::Parser::new();
    match ext {
        "rs" => parser
            .set_language(&tree_sitter_rust::LANGUAGE.into())
            .expect("Failed to set Rust language"),
        "js" | "ts" | "jsx" | "tsx" => parser
            .set_language(&tree_sitter_javascript::LANGUAGE.into())
            .expect("Failed to set JS language"),
        "py" => parser
            .set_language(&tree_sitter_python::LANGUAGE.into())
            .expect("Failed to set Python language"),
        _ => {
            eprintln!("Unsupported extension: {ext}");
            std::process::exit(1);
        }
    }

    // Warmup parse
    let _ = parser.parse(&source, None).expect("Parse failed");

    // Timed iterations
    let start = Instant::now();
    let mut node_count = 0u64;
    for _ in 0..iterations {
        let tree = parser.parse(&source, None).expect("Parse failed");
        // Walk all nodes to ensure full parse
        let mut cursor = tree.walk();
        let mut stack = vec![true]; // true = first visit
        loop {
            if stack.last().copied() == Some(true) {
                node_count += 1;
                *stack.last_mut().unwrap() = false;
                if cursor.goto_first_child() {
                    stack.push(true);
                    continue;
                }
            }
            if cursor.goto_next_sibling() {
                *stack.last_mut().unwrap() = true;
                continue;
            }
            stack.pop();
            if !cursor.goto_parent() {
                break;
            }
        }
    }
    let elapsed = start.elapsed();

    let avg_ms = elapsed.as_secs_f64() * 1000.0 / iterations as f64;
    let total_ms = elapsed.as_secs_f64() * 1000.0;

    println!("{{");
    println!("  \"tool\": \"rust\",");
    println!("  \"operation\": \"tree_sitter_parse\",");
    println!("  \"file\": \"{file_path}\",");
    println!("  \"lines\": {line_count},");
    println!("  \"bytes\": {},", source.len());
    println!("  \"iterations\": {iterations},");
    println!("  \"total_ms\": {total_ms:.2},");
    println!("  \"avg_ms\": {avg_ms:.3},");
    println!("  \"nodes_per_iter\": {}", node_count / iterations as u64);
    println!("}}");
}
